### ISSUE TITLE
RFC: Splitting off special function scalar kernels into separate library

### ISSUE THREAD

**Issue by steppi** [scipy/scipy#21649]
Something that's been under discussion for a while now is splitting off the scalar kernels for ufuncs and gufuncs in `scipy.special` into a separate library. The plan would be for this library to still be managed under the `scipy` organization, but it would contain GPU capable implementations of scalar kernels which could be used in other array libraries, and would be a shared dependency of at least SciPy and CuPy.

As a start, I've created the [scipy/xsf](https://github.com/scipy/xsf) repo for this purpose with the blessing of @rgommers. I wrote up a longish RFC at https://github.com/scipy/xsf/issues/1. This gives some motivation, describes the migration steps I anticipate, and gives a rough roadmap for the rest of the year. If we can get consensus for this, I'd like to complete the migration over the next month.

My intention is for this issue in SciPy as a place to discuss whether it's a good idea to split out the scalar kernels in general, and what this may mean for SciPy, while https://github.com/scipy/xsf/issues/1 is for discussing specifics of the migration plan and roadmap. I don't mean for this to be strict though; feel free to comment on either.

**Comment by lucascolley** [comment]
Not to try to add yet another place for people to comment but probably good to post about this on Discourse too.

**Comment by steppi** [comment]
> Not to try to add yet another place for people to comment but probably good to post about this on Discourse too.

I'm typing the discourse post now.

**Comment by tylerjereddy** [comment]
I think my main concern from the discussion over there is:

> ... we don't need to test xsf in SciPy which would be major win for SciPy and whoever adopts xsf along the way.

I think I see why abstraction of `xsf` seems appealing in general with reduced duplication of effort across the ecosystem and greater freedom to do super-detailed testing/development in a purpose-built external library/testsuite for the `special` machinery. There may also be "social" advantages if you get the `special` function developers to convene around a single source of truth/repo and have fruitful discussions with each other instead of being overwhelmed with their individual maintenance assignments.

That said, I don't think most decisions can be made without at least a few drawbacks--it is hard to beat the convenience of having the regression test and its patch applied directly in the same repo right away. You can argue that we only really need to sniff test the wrappers, but for example--maybe a bug comes up related to the parent-level SciPy `meson` build flags we are using to compile the submodule headers. Our testsuite may not catch it if we only do superficial `special` wrapper testing, and it may not be a problem that can be addressed in the `xsf` testsuite because there is no numerical correctness issue if the build/link flags are fixed on our end.

So I think the main thing I'd like to see is some leeway in allowing us to add regression tests as we see fit--not getting too aggressive about banning them in SciPy, even if the `hypothesis`-level super-crazy theoretical testing of near-singularity behavior happens in `xsf` proper I think SciPy maintainers should still have fairly broad discretion in adding `special` tests in the main repo if they think it will prevent regressions.

**Comment by ilayn** [comment]
@tylerjereddy I think you meant to type this in the other rfc? But I agree with your comment wherever you want to keep it (here or https://github.com/scipy/xsf/issues/1). I think we can find a middle ground that we don't really test all the way to the ULP correctness. But we also do not make a full suite that makes our CI blocked for 8 minutes.

**Comment by tylerjereddy** [comment]
I did mean to put it here I think, as a main repo concern; ok, sounds like we agree then, which makes this not much different than testing other things we vendor/wrap.

**Comment by steppi** [comment]
I also agree with you @tylerjereddy. It's important to strike a balance and not just rely on a blind rule for what kind of tests should go where. By placing exhaustive numerical tests in `xsf` and leaving them out of SciPy, we can cut down on SciPy's test times, but it's still important for SciPy to have less exhaustive regression tests to cover things like you mentioned.

**Comment by rgommers** [comment]
Linking the Discourse post: https://discuss.scientific-python.org/t/splitting-off-scipys-special-function-scalar-kernels-into-a-new-library/1468. 4 hearts, no replies. 

Looks like we're ready to finish off this move.
